Add Performance Points & Star Rating updates newspost/assets (#5890)
ppy/osu-wiki: news/2021-07-26-performance-points-star-rating-updates.md

---
layout: post
title: "Performance Points & Star Rating Updates"
date: 2021-07-26 10:00:00 +0000
---

A brand new set of changes to the performance points and star rating algorithms have just gone live - read on to discover what's new, and where things are headed in the future!

![](/wiki/shared/news/2021-07-27-performance-points-star-rating-updates/star-rating.jpg)

Thanks to the efforts of some very bright minds, a new batch of changes are live for both the osu! and osu!mania gamemodes. Read on below to check out all the changes stated in full (courtesy of a write-up from [spaceman_atlas](https://osu.ppy.sh/users/3035836)), including a number of graphs and charts to help things along.

Please note: the current batch of changes contain things that will take effect immediately, as well as ones which will have no visible change for the moment but will lay out the groundwork for improvements in the future.

> Additionally, the changes below are split into "star rating" and "performance points" sections, but do note that anything that affects star rating will also impact performance points, as both are based off an algorithmic *skill* value which is dynamically generated from the beatmap being played.

Reviewer: ```suggestion
Additionally, the changes below are split into "star rating" and "performance points" sections, but do note that anything that affects star rating will also impact performance points, as both are based off algorithmic *skill* values which are dynamically generated from the beatmap being played.
```

That's where [**emu1337**](https://osu.ppy.sh/users/2185987) stepped in, and proposed [a change](https://github.com/ppy/osu/pull/13483) that applies a damping factor to the 10 most difficult sections for the aim skill and the 5 most difficult for the speed skill.

This means that the final star rating values will decrease across the board for those skills, but maps with high but short strain spikes will be impacted more than maps with consistent strain values throughout.

You can take a peek at the following graphs to check out how this looks in practice:

![](/wiki/shared/news/2021-07-27-performance-points-star-rating-updates/weighting_comparison_1.png)

![](/wiki/shared/news/2021-07-27-performance-points-star-rating-updates/weighting_comparison_2.png)

A couple of examples of difficulties that will be impacted by this change more than others are [Brian the Sun - Lonely Go! (TV Size) \[Fiery's Extreme\]](https://osu.ppy.sh/beatmapsets/863227#osu/1860433) and [MOMOIRO CLOVER Z - SANTA SAN \[1-2-SANTA\]](https://osu.ppy.sh/beatmapsets/1045600#osu/2185824).

> ### All rulesets: Fixed strain skills incorrectly reading out time values when rate was modified by mods

Reviewer 2: @smoogipoo does this mean we need to run recalculations for taiko/catch as well..?

Reviewer 3: All rulesets have changed slightly, so ideally we'd run for all rulesets. I'll try to find out exactly how much taiko/catch have changed.

Reviewer 3: List of taiko/catch changes >0.1*: https://docs.google.com/spreadsheets/d/17O7Eq-DBH_Aw2cGX3k4JWEIMRsStm9Qb5zNxL5c9Vis/edit?usp=sharing

I'd do a reprocess of both anyway.

### osu!: Approach rate and Flashlight bonuses are now mutually exclusive

As mentioned previously, approach rate values between 8 and 10.33 are known to give a bonus to a given play. When playing with the Flashlight mod enabled, another bonus was also applied.

Previously, these values stacked together in the end result, which makes very little sense given that approach circles become increasingly less visible when Flashlight ramps up at higher combos.

Going forward, only the larger of the two bonuses will be applied in a [set of changes](https://github.com/ppy/osu/pull/13483) laid out by [**emu1337**](https://osu.ppy.sh/users/2185987).

Additionally, these changes slightly reduce aim skill rewards in the case of wide-angle jumps. See the graph below for a rough visual that aims to convey how the change will impact jumps:

![](/wiki/shared/news/2021-07-27-performance-points-star-rating-updates/jump_angle_bonus.png)

A few notes as to how to interpret the plot:

- The angular axis represents the angle between two consecutive movements performed by a player who is jumping between three hitcircles.

> - The radial axis represents the normalised distance between the hitcircles.

Reviewer 4: I might have reached a bit far with "angular axis"... It does not seem to actually be a thing indeed.

  For simplicity the graph assumes that the two movements are of the same length (but the real implementation also factors in differing distances).

> - The bonus also has a timing-based component not shown above. On the plot above it is assumed that the timing factor is minimal (and so the angle bonus is the biggest it can be)

Reviewer 4: Not sure if I agree with splitting that first bullet off, it's meant to be a follow-up to the previous sentence about what the radius is.

> - The bonus also has a timing-based component not shown above. On the plot above it is assumed that the timing factor is minimal (and so the angle bonus is the biggest it can be)

Reviewer: This sentence doesn't mention the radius, nor does it explicitly explain what the radius is (the first sentence does that). Besides, this sentence is also a point on "how to interpret the plot".

Though solely in the domain of osu!lazer, it may eventually resolve a [fair](https://github.com/ppy/osu-web/issues/7846) [few](https://github.com/ppy/osu-web/issues/6075) [issues](https://github.com/ppy/osu-web/issues/5442) with displaying mania map difficulties on the website, by reusing lazer difficulty calculation components in other areas of the osu! infrastructure.

### All rulesets: More flexibility in the structure of skills

As community members explore new ideas in the domain of difficulty and performance calculation, the limitations of how things were previously set up can come to bear in quite stifling ways. 

Long-term contributor [**Syrin**](https://osu.ppy.sh/users/5701575) has made great strides in revamping this process for the benefit of anyone interested in creating their own rulesets in the changes listed below:

- [Skills now receive a list of active mods](https://github.com/ppy/osu/pull/11687), which will allow their creators to very easily create mod-specific adjustments in the future. Notably, this new capability has already been utilized to [clean up existing osu!catch star rating calculations](https://github.com/ppy/osu/pull/11857).

> - Skills were previously hard limited in how many objects they could operate on. [A new change](https://github.com/ppy/osu/pull/11688) allows rules to now manually specify directly how many objects they can act on in relation to any one object, allowing for much more robust algorithms to be developed.

Reviewer: in that case it would be

```suggestion
- Skills were previously hard-limited in how many objects they could operate on. [A new change](https://github.com/ppy/osu/pull/11688) now allows direct and manual specification of how many objects can be acted on in relation to any one object, allowing for much more robust algorithms to be developed.
```

You may also wish to consult the [performance points](/wiki/Performance_points) and [star rating](/wiki/Beatmapping/Star_rating) articles to broaden your understanding of these surprisingly complicated topics.

As a brief reminder/summary, the following values are a core part of the performance points algorithm:

- **Aim**, which considers how difficult it is to consistently hit notes back-to-back during play (affected by things like approach rate, jumps, Flashlight, Hidden, Hard Rock and more)
- **Speed**, which considers how quickly new elements are presented to the player during the course of the map (affected by things like the number of hit objects over time, BPM, Double Time and Half Time)
- **Accuracy**, which considers the player's ability to stay on time rhythmically and how difficult this is to achieve (affected by things like Hidden, Hard Rock, Flashlight and more)
- **Strain**, which considers how often and how long a player has to endure intense or technically demanding patterns during play (affected by things like streams, fast jumps and more)

## Release Schedule

> Deployment of these changes will as usual, take some time to propagate out. As this batch of changes affects all game modes, it is likely to take longer than usual. If you are interested in following the progress we recommend you bookmark this page and check back on this section, as it will be updated as forward progress is made.

Reviewer: the deployment will propagate or the changes will propagate? (maybe doesn't matter, but this reads better)

```suggestion
Deployment of these changes will as usual take some time, but since this batch of changes affects all game modes, they will likely take longer than usual to propagate. If you are interested in following the progress we recommend you bookmark this page and check back on this section, as it will be updated when forward progress is made.
```

Reviewer 2: Are you sure about "when" instead of "as"?

Reviewer: deliberated on this a bit, because they mean the same thing in practice, but i guess it's fine to keep the emphasis on the gradual-ness

### 🏃 Star rating updates \[2021-07-28\]

An initial pass over all beatmaps to update their star ratings. This will take immediate effect for all newly submitted scores.

Estimated to take less than one day.

### 🛑 Background user total recalculations \[2021-07-29\]

Starting from the user at rank #1 and working outwards, each user's total pp will be recalculated.

> During this period, rank history graphs will be frozen. If a user plays, their **total PP** value will immediately be updated using the correct values for all past scores, but this won't match past scores' display under "Best Performance". As a result, it may seem as if you temporarily gain, lose or fluctuate in rank relative to others.

Reviewer: probably fine in lowercase, but feel free to change that

```suggestion
During this period, rank history graphs will be frozen. If a user plays, their **total pp** value will immediately be updated using the correct values for all past scores, but this won't match past scores' display under "Best Performance". As a result, it may seem as if you temporarily gain, lose or fluctuate in rank relative to others.
```

- As part of this maintenance, we will delist any inactive users (anyone that has not played the game in 3 months).
- Scores in "best performance" may look to be out of order or not visible at this point.

Estimated to take 2 days for the majority of active users, and longer for lower ranking users. The initial run is targeting all users above 1,000 total pp, after which we will begin to start a longer running update that may run over a week or so.

### 🛑 Re-enabling of global rank history updates \[TBD\]

Rank history graphs updates will be enabled and updated again. At this point, all users' global leaderboard ranks will be stable going forward.

### 🛑 Reindexing \[TBD\]

This will fix scores being out of order on profiles, and in some rare cases not being displayed at all.

Estimated to take 2-4 days.

### 🛑 Background user score recalculations \[TBD\]

> While the above "user total" step is aimed to cover the majority of visible scores in profile "Best Performance" section, this is a second pass operation which aims to address all remaining scores that may have not been updated yet.

Reviewer: "the above" usually refers to something just mentioned

```suggestion
While the "user total" step from before is aimed to cover the majority of visible scores in the "Best Performance" profile section, this is a second-pass-operation which aims to address all remaining scores that may have not been updated yet.
```

Reviewer 2: "aforementioned" maybe?

Reviewer: edited
